### PR TITLE
Optimizations for field arithmetic

### PR DESCRIPTION
**Pull request by henrydcase**
Commit 1:  perf p503: improve performance by using branch predictor

```       perf p503: improve performance by using branch predictor
    
    In case of x86, there are few implementation of the montgomery reduction
    and multiplication. At runtime, library chooses most performant
    implementation according to information received from CPUID. The
    implementation then is assigned to a function pointer which gets CALL'd
    during program execution.
    
    The problem is a mixture of following; function pointer points to an
    assembly function, it's arguments are also pointers to some data and
    arguments passed to the call are allocated on a stack before a call.
    
    As variable can't be tagged as go:noescape, the go compiler will move
    arguments passed to the call from stack to heap. This causes significant
    performance degradataion.
    
    The solution is not to use function pointer. Instead, both redc and mul,
    will check at runtime CPU capabilities and do the JMP to the correct
    part of the text. Thanks to branch prediction the cost of the solution
    is minimal and smaller than function call. This patch also removes all
    heap allocations done by functions operating on prime field.
    
    The other goal of this patch is to remove x86 specific code from
    arith_decl.go, which will be also compiled for ARM arch in near future.
    
    Results:
    benchmark                            old ns/op     new ns/op     delta
    BenchmarkFp2ElementMul               428           194           -54.67%
    BenchmarkFp2ElementInv               67353         34447         -48.86%
    BenchmarkFp2ElementSquare            335           139           -58.51%
    BenchmarkFp503MontgomeryReduce       26.3          22.8          -13.31%
    BenchmarkSidhKeyAgreementP503        12451199      6402396       -48.58%
    BenchmarkAliceKeyGenPubP503          7349333       3590954       -51.14%
    BenchmarkBobKeyGenPubP503            8253676       4094141       -50.40%
    BenchmarkSharedSecretAliceP503       5888022       2916821       -50.46%
    BenchmarkSharedSecretBobP503         6908018       3436713       -50.25%
    
    Comparision with P751:
    ----------------------
    BenchmarkBobKeyGenPubP751           13616876
    BenchmarkBobKeyGenPubP503            4094141
    BenchmarkSharedSecretAliceP751       9870216
    BenchmarkSharedSecretAliceP503       2916821

    There is a cost - possibly CMPB & JE each time mul and redc is called.
    Also patch introduces arith_amd64.go file which keeps some variables -
    this currently needs to be done for each field. Probably it could be
    possible to get rid of it at some point.
```

Commit 2: similar change for P751
```
    BenchmarkSidhKeyAgreementP751      23839233      20241485      -15.09%
    BenchmarkAliceKeyGenPubP751        13733691      11180516      -18.59%
    BenchmarkBobKeyGenPubP751          14804459      12889797      -12.93%
    BenchmarkSharedSecretAliceP751     10649209      9155030       -14.03%
    BenchmarkSharedSecretBobP751       12739399      10751825      -15.60%
```

### CODE REVIEW

File: p503/arith_amd64_test.go [hunk 1]
```diff
@@ -43,32 +73,56 @@ func testRedc(t *testing.T, redcFunc func(z *FpElement, x *FpElementX2)) {
 
 // Ensures corretness of implementation of mul operation which uses MULX
```
Comment (Ko-) [inline p503/arith_amd64_test.go]: Typo: correctness. (6 times)

File: p503/arith_amd64_test.go [hunk 2]
```diff
@@ -43,32 +73,56 @@ func testRedc(t *testing.T, redcFunc func(z *FpElement, x *FpElementX2)) {
...
+	defer recognizecpu()
 	if !cpu.HasBMI2 {
 		t.Skip("MULX not supported by the platform")
 	}
-	testRedc(t, redcWithMULX)
+	testRedc(t, kUse_MULX, kUse_MUL)
 }
 
 // Ensures corretness of Montgomery reduction implementation which uses MULX
+// and ADX
```
Comment (Ko-) [inline p503/arith_amd64_test.go]: We're not really being consistent with this. But BMI2 is the instruction set extension that adds the MULX instruction. And ADX is the instruction set extension that contains the ADCX and ADOX instructions.  Here it's about the instructions, so it should say ADOX/ADCX.

This is also a bit confusing in `cpuid.go` and in  a number of labels.

File: p503/arith_amd64_test.go [hunk 3]
```diff
@@ -43,32 +73,56 @@ func testRedc(t *testing.T, redcFunc func(z *FpElement, x *FpElementX2)) {
...
 func TestRedcWithMULXADX(t *testing.T) {
+	defer recognizecpu()
+	if !(cpu.HasADX && cpu.HasBMI2) {
+		t.Skip("MULX, ADCX and ADOX not supported by the platform")
+	}
+	testRedc(t, kUse_MULXADX, kUse_MUL)
+}
+
+// Ensures corretness of Montgomery reduction implementation which uses MULX
+// and ADX.
```
Comment (Ko-) [inline p503/arith_amd64_test.go]: Idem.

File: p751/arith_amd64.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
+// +build amd64,!noasm
+
+package p751
+
+import cpu "github.com/cloudflare/p751sidh/internal/utils"
+
+// There couple of reasons for having those variables here:
+// * to have an access to them from assembly
+//   TODO(kk): Is there a way to access variable from different package?
```
Comment (Ko-) [inline p751/arith_amd64.go]: See p503. Let's remove these files.

File: p751/arith_amd64.s [hunk 5]
```diff
@@ -2379,7 +2345,20 @@ TEXT ·fp751MontgomeryReduceFallback(SB), $0-16
 	ADCQ	$0, R10
 	ADDQ	(184)(REG_P1), R10		// Z11
 	MOVQ	R10, (88)(REG_P2)		// Z11
+	RET
+
+redc_with_mulx_adx:
```
Comment (Ko-) [inline p751/arith_amd64.s]: `redc_with_bmi2_adx`?

File: p751/arith_amd64.s [hunk 6]
```diff
@@ -2379,7 +2345,20 @@ TEXT ·fp751MontgomeryReduceFallback(SB), $0-16
...
+	// This implements the Montgomery reduction algorithm described in
+	// section 5.2.3 of https://eprint.iacr.org/2017/1015.pdf.
+	// This assumes that the BMI2 and ADX instruction set extensions are available.
+	REDC(0(REG_P2), 0(REG_P1), mul256x448bmi2adx)
+	RET
 
+redc_with_mulx:
```
Comment (Ko-) [inline p751/arith_amd64.s]: `redc_with_bmi2`?

File: p503/arith_amd64_test.go (hunk 1 above)
Comment (henrydcase) [reply p503/arith_amd64_test.go]: correctness written incorrectly ...

File: p751/arith_amd64.s (hunk 5 above)
Comment (henrydcase) [reply p751/arith_amd64.s]: I prefer ``redc_with_mulx_adcx_adox`` as those macros don't use whole instruction set, but specific instructions

**Pull request merged**